Test checkpoint validity before starting simulation

roelof-groenewald opened the pull request.

roelof-groenewald committed:
build image - test for non-corrupt checkpoints before starting simulation

KZhu-ME commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -142,7 +142,21 @@ def run_restart(checkpoint_directory="diags",
             return False, None, None
 
-    checkpoint = checkpoints[-1]
+    # check if the checkpoint was completely written by testing if a flux diag
```
Is there a case where the checkpoint file could be saved if it was on a diag step and we could copy over the flux diag file from diags to the checkpoint? I'd just like to avoid automatically deleting files unless the file is absolutely unusable.

Also I think it would make more sense for this check to be put earlier, when the program checks for valid checkpoints, and errors out if none are found. The check could also print a warning telling the user that no fluxdiag was found for this checkpoint file, and that it would be using an earlier checkpoint instead.

roelof-groenewald replied:
On the first point: We have no way of knowing (that I'm aware of) whether the WarpX checkpoint is valid until we attempt a restart and see WarpX fail (usually with something like `terminate called after throwing an instance of 'std::__ios_failure'`). My opinion is that we are better of playing it safe - only attempt restarts from files we know are definitely good - rather than risking having jobs fail. If there was a good way to test the validity of a checkpoint I would agree with you.

On the second point: Good idea, I'll move the check into `get_sorted_checkpoints`.

KZhu-ME replied:
That makes sense. In that case, what do you think about tying the deletion of the checkpoint files with the `clear_old_checkpoints` flag, so automatic removal of checkpoints can be turned on or off at the user's discretion? If the flag is not set, the program can just skip over the checkpoint that is missing the `fluxdata` file, and move to the next latest checkpoint with a logged warning.

peterscherpelz replied:
At least move to a corrupted subdir, so let's just leave it as deleted by default: And we can add this later if necessary for debugging.

roelof-groenewald replied:
@KZhu-ME I decided not to move the new functionality in `get_sorted_checkpoints` since that way I can reuse it for `_clean_corrupt_checkpoints`.

roelof-groenewald committed:
build image - only delete checkpoints if a restart was successful from an earlier checkpoint

roelof-groenewald commented:
@PhilMiller We have a `force_restart` flag that allows the user to set up the simulation to error out if it cannot find a valid restart file. With this functionality already in place I don't think it is still necessary to only check the two most recent checkpoint files for validity.

roelof-groenewald committed:
only clean possibly corrupt checkpoints when restarting

peterscherpelz reviewed (changes requested):
I'm still uneasy about this. The basic behavior is OK but it's so brittle against edge cases I'm worried something bad will happen. Example: Someone actually moves rather than copies fluxdiag files from a run that's been going for a month. Desired behavior: There would be two checkpoint always present. At most the code would remove one, and then upon failing to restart from the other one, realize something is up and not delete it. Or not delete either.

Instead, right now, in the case above, it'll return `False` for the restart flag, which will start the run from the beginning, but then it won't trigger `_clean_corrupted_checkpoints` because `self.restart` is False. Although this is desired behavior in my edge case, it's not desired behavior when the very first checkpoint is indeed corrupted - it shouldn't stick around just because the simulation is starting fresh.

peterscherpelz committed:
Revise handling of corrupted checkpoints.

This attempts to be relatively cautious/fail-safe by at most removing
one corrupt checkpoint (and two are always kept by default so removing
one will leave another), and erroring out if two or more appear to be
corrupt.

roelof-groenewald commented:
@peterscherpelz Thanks for the changes. I'm happy to keep that approach over mine. Have you tested this yet on the various edge cases?

One thing to note is that when a new run is started we always delete the `diags` folder, so when `self.restart` is False (and `_clean_corrupted_checkpoints` was not run in the old way) all checkpoints are deleted anyway.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -61,15 +61,16 @@ def get_sorted_checkpoints(checkpoint_directory, checkpoint_prefix):
 def clean_old_checkpoints(checkpoint_directory="diags",
                           checkpoint_prefix=DEFAULT_CHECKPOINT_NAME,
-                          num_to_keep=1):
+                          num_to_keep=2):
```
I think we can still keep this as 1 since the old checkpoint is only deleted once the new one is completely written anyway. The new approach doesn't specifically need 2 checkpoints, it just needs at least 1 valid one.
```suggestion
                          num_to_keep=1):
```

peterscherpelz replied:
Yeah, but this falls into my concern about it being cautious/fail-safe rather than fail-bad: If we only keep 1 checkpoint, and then say (to use the same example) someone accidentally moves the fluxdata.ckpt file, that 1 checkpoint will be deleted on restart because it's assumed corrupt, and then you're done for - you've lost the whole run progress no matter what. By keeping 2, it makes the whole system robust against any of those weird circumstances where we don't want to auto-delete a checkpoint.

PhilMiller replied:
Can you give me time to look over this tomorrow afternoon before merging it? I can probably identify a lot of other returnable edge cases, and figure out whether they're covered

PhilMiller replied:
I just looked at where this is getting called, in `CheckPointDiagnostic`. That keeps its own member `num_to_keep` which it passes to this. It defaults to 1. I think that's a serious problem, because the 1 that it will keep will be the one that it just wrote. Since that write may not be immediately durable, it's definitely worth keeping its predecessor (i.e. at least 2). Given that, let's fix that, and also leave the default argument here at 2.

PhilMiller commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -89,6 +90,98 @@ def clean_old_checkpoints(checkpoint_directory="diags",
         shutil.rmtree(dirpath)
 
 
+def _eval_corrupt_checkpoint(checkpoint_dir, checkpoint):
+    """Determine if a checkpoint appears to be corrupt by checking if
+    fluxdata.ckpt is present.
```
```suggestion
def _eval_checkpoint_validity(checkpoint_dir, checkpoint):
    """Determine if a checkpoint appears to be valid by checking if
    fluxdata.ckpt is present.
```

Almost always better to phrase names in the affirmative, especially since the function is actually returning the affirmative.

PhilMiller commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -89,6 +90,98 @@ def clean_old_checkpoints(checkpoint_directory="diags",
+    # list.
+    if len(checkpoint_list) == 1:
+        _remove_corrupt_checkpoint(checkpoint_dir, checkpoint_list[-1])
+        return []
```
There's an edge case here of jobs never making progress, because they're killed too soon. That's a meta-issue to this PR, though, since there's nothing we can do about it directly at this level.

Rather, we may want to keep a count of whether this case is occurring repeatedly, and either abort the run or automatically shorten the checkpoint interval to compensate

Does AWS expose a variable for which 'attempt' at a job a given execution represents?

KZhu-ME replied:
`AWS_BATCH_JOB_ATTEMPT` environment variable tracks the retry number. https://docs.aws.amazon.com/batch/latest/userguide/job_retries.html

PhilMiller commented:
I think the logic here, as revised in response to Peter's comments, is ok. I have some ideas for hardening the code that writes out the checkpoints, to help avoid some potential edge cases.

roelof-groenewald committed:
suggested changes from PM code review

roelof-groenewald committed:
Merge branch 'upstream_update' into rgdevel

roelof-groenewald committed:
updated version and changelog
